File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <iostream>
#include <queue>
#define max(a,b) ((a) > (b) ? (a) : (b))
#define min(a,b) ((a) > (b) ? (b) : (a))
#define pr pair<int,int>
#define x first
#define y second

using namespace std;

int n,m;

/*
void BFS(queue<POS> &q, POS p, vector<vector<int> > maps)
{
    if(maps[min(p.x+1,m)][p.y] == 1)
    {
        q.push({p.x+1,p.y, p.level+1});   
    }    

    if(maps[max(0, p.x-1)][p.y] == 1)
	{
		q.push({p.x-1,p.y,p.level+1});
	}
    
    if(maps[p.x][min(p.y+1, n)] == 1)
	{
		q.push({p.x,p.y+1,p.level+1});
	}
    
    if(maps[p.x][max(0, p.y-1)] == 1)
	{
		q.push({p.x,p.y-1,p.level+1});
	}
}*/
int solution(vector<vector<int> > maps)
{
    queue<pr> q;
    q.push({0,0});
    m = maps.size()-1;
    n = maps[0].size()-1;
    int **dp = new int*[maps.size()];
    for(int i=0;i<maps.size();i++)
        dp[i] = new int[maps[0].size()];
    
    dp[0][0] = 1;
    maps[0][0] = 0;
    while(q.size()>0)
    {
        pr pos = q.front();
        if(pos.x == m && pos.y == n)
            return dp[pos.x][pos.y];
        
        q.pop();
        //maps[pos.x][pos.y] = 0;
        
        if(maps[min(pos.x+1,m)][pos.y] == 1)
        {
            maps[min(pos.x+1,m)][pos.y] = 0;
            q.push({pos.x+1,pos.y});   
            dp[pos.x+1][pos.y] = dp[pos.x][pos.y] + 1;
        }    

        if(maps[max(0, pos.x-1)][pos.y] == 1)
	    {
            maps[max(0, pos.x-1)][pos.y] =0;
		    q.push({pos.x-1,pos.y});
            dp[pos.x-1][pos.y] = dp[pos.x][pos.y] + 1;
	    }
    
        if(maps[pos.x][min(pos.y+1, n)] == 1)
	    {
            maps[pos.x][min(pos.y+1, n)] = 0;
	    	q.push({pos.x,pos.y+1});
            dp[pos.x][pos.y+1] = dp[pos.x][pos.y] + 1;
	    }
    
        if(maps[pos.x][max(0, pos.y-1)] == 1)
	    {
            maps[pos.x][max(0, pos.y-1)] = 0;
		    q.push({pos.x,pos.y-1});
            dp[pos.x][pos.y-1] = dp[pos.x][pos.y] + 1;
	    }
    }
    return -1;
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.io.*;
import java.util.*;

class Solution {
    int targetX; int targetY;
    int [][]map; int answer=99999999;
    int []dx = {1, 0, -1, 0};
    int []dy = {0, -1, 0, 1};
    void bfs(int x, int y){
        // if(map[x][y] != 1)  return ;
        // map[x][y] = 1;
        Queue<Pair> que = new LinkedList<>();
        que.add(new Pair(0,0));

        while(!que.isEmpty()){
            Pair pair = que.poll();
            x = pair.x; y = pair.y;
            if(pair.x==targetX && pair.y==targetY) {
                answer = Math.min(map[x][y],answer);
                return;
            }
            for(int i=0;i<4;i++){
                int cur_x = x+dx[i];    int cur_y = y+dy[i];
                if(cur_x <0 || cur_x > targetX || cur_y <0 || cur_y > targetY) continue; // 좌표 벗어난 경우
                if(map[cur_x][cur_y] !=1 ) continue; // 1로만 잡으면 한번 갔던길을 다시 갈 수 없음///
                // if(map[cur_x][cur_y] <= map[x][y] +1) map[cur_x][cur_y] = map[x][y] +1;
                map[cur_x][cur_y] = map[x][y] +1;

                que.add(new Pair(cur_x, cur_y));
                
            }
        }
    }
    
    
    public int solution(int[][] maps) {
        map = maps;
        targetX = maps.length-1;
        targetY = maps[0].length-1;
        bfs(0,0);
        
        if(answer == 99999999) answer = -1;
        return answer;
    }
}


class Pair{
    int x; int y;
    public Pair(int x, int y){
        this.x = x;
        this.y = y;
    }
}
